File: 20-actors/watari/methods/kotoba.py
#!/usr/bin/env python3
"""kotoba.py — watari kotoba Datom-log writer (local, content-addressed). ADR-2606041827
+ ADR-2605262130 + ADR-2605312345.

The substrate boundary (root CLAUDE.md): canonical state is the **kotoba Datom log** —
content-addressed EAVT assertions, append-only (非終末論; the fix stream IS the trajectory, the
latest fix IS the current position). watari had a graph + analyzer + an R0 ingest stub but no
self-driving loop and no local log; this module is the **local, autonomous-loop** write path — the
same path shionome / ipaddress / yabai / sukashi / watatsuna use (`methods/autorun.py`): a
self-driving heartbeat appends content-addressed transactions to a local append-only EDN log with
NO external I/O, so watari can run its own observe→analyze→persist live-craft-situation cycle on
the Murakumo fleet without a human or a live node in the loop.

Constitutional posture is preserved by construction (watari hard rules): outputs are
situational-awareness, not surveillance / not targeting (G2 — aggregate lane/chokepoint/approach
density); a craft is a craft, NEVER a person — no position track is linked to a named individual,
no pattern-of-life (G4, the defining gate; operator is a company only). The loop persists exactly
what `analyze.py` already computes, with derived signals flagged :movement/derived.

  - graph_datoms(rows)         → EAVT assertions for every entity (craft / fix / leg / lane).
  - derived_datoms(craft, ln, a) → EAVT assertions for the analyzer's derived :movement/* signals
                                   (chokepoint-transit, lane-load by kind, stale-track tail),
                                   flagged :movement/derived.
  - make_tx / append_tx / read_log / head_cid / verify_chain — content-addressed commit-DAG.

EAVT = [op entity attribute value]; op is :db/add only (append-only — no :db/retract). Stdlib only.
Deterministic: the caller supplies tx_id + as_of (no wall clock) → resume-safe.
"""
from __future__ import annotations

import hashlib
import json
import os
import pathlib
import sys
from typing import Any

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from analyze import _parse, _tokens  # noqa: E402  (the inline EDN reader lives in analyze.py)

LOG_DEFAULT = (pathlib.Path(__file__).resolve().parents[1] / "data"
               / "watari.datoms.kotoba.edn")

ID_KEYS = (":craft/id", ":craft.fix/id", ":craft.leg/id", ":lane/id")


def _add(entity: str, attr: str, value: Any) -> list:
    """One append-only EAVT assertion: [:db/add <entity> <attr> <value>]."""
    return [":db/add", entity, attr, value]


def graph_datoms(rows: list) -> list[list]:
    """Flatten the moving-craft graph into append-only EAVT assertions. E = the entity's id;
    cardinality-many list values fan out. Persists craft identity / position fixes / legs / lanes
    as-is — the seed carries operator-as-company only, never a person (G4)."""
    out: list[list] = []
    for r in rows:
        if not isinstance(r, dict):
            continue
        e = next((r[k] for k in ID_KEYS if k in r), None)
        if e is None:
            continue
        for k, v in r.items():
            if k in ID_KEYS:
                continue
            for item in (v if isinstance(v, list) else [v]):
                out.append(_add(e, k, item))
    return out


def derived_datoms(craft: dict, lanes: dict, a: dict) -> list[list]:
    """Flatten the analyzer's derived :movement/* signals into EAVT assertions, each flagged
    :movement/derived true (aggregate situational-awareness recomputed on read, never re-ingested
    as fact, never a per-craft follow/targeting feed — G2/G4). Mirrors analyze.render_datoms.
    `a` is analyze.analyze()."""
    out: list[list] = []
    for cp in sorted(a["choke_transit"], key=lambda k: -a["choke_transit"][k]):
        e = f"movement-choke-{cp}"
        out += [_add(e, ":movement/chokepoint", cp),
                _add(e, ":movement/chokepoint-transit", a["choke_transit"][cp]),
                _add(e, ":movement/derived", True)]
    for ln in sorted(a["lane_load"], key=lambda k: -a["lane_load"][k]):
        vk = len(a["lane_kind"][ln].get(":vessel", set()))
        ak = len(a["lane_kind"][ln].get(":aircraft", set()))
        e = f"movement-lane-{ln}"
        out += [_add(e, ":movement/lane", ln),
                _add(e, ":movement/lane-load", a["lane_load"][ln]),
                _add(e, ":movement/vessels", vk),
                _add(e, ":movement/aircraft", ak),
                _add(e, ":movement/derived", True)]
    for c in sorted(a["stale"]):
        fx = a["latest"][c]
        e = f"movement-stale-{c}"
        out += [_add(e, ":movement/craft", c),
                _add(e, ":movement/stale", True),
                _add(e, ":movement/last-seen", fx.get(":craft.fix/observed-at", "")),
                _add(e, ":movement/derived", True)]
    return out


def _canonical(datoms: list[list], prev_cid: str) -> bytes:
    return json.dumps({"prev": prev_cid, "datoms": datoms},
                      ensure_ascii=False, sort_keys=True, separators=(",", ":")).encode("utf-8")


def tx_cid(datoms: list[list], prev_cid: str = "") -> str:
    """Content address = sha256 over (prev_cid, datoms) → a commit-DAG."""
    return "b" + hashlib.sha256(_canonical(datoms, prev_cid)).hexdigest()


def make_tx(datoms: list[list], *, tx_id: int, as_of: int, prev_cid: str = "") -> dict:
    return {
        ":tx/id": tx_id,
        ":tx/as-of": as_of,
        ":tx/prev": prev_cid,
        ":tx/cid": tx_cid(datoms, prev_cid),
        ":tx/count": len(datoms),
        ":tx/datoms": datoms,
    }


def _edn_val(v: Any) -> str:
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, (int, float)):
        return repr(v)
    if isinstance(v, str):
        return v if v.startswith(":") else json.dumps(v, ensure_ascii=False)
    if isinstance(v, list):
        return "[" + " ".join(_edn_val(x) for x in v) + "]"
    return json.dumps(str(v), ensure_ascii=False)


def _tx_to_edn(tx: dict) -> str:
    datoms = " ".join("[" + " ".join(_edn_val(x) for x in d) + "]" for d in tx[":tx/datoms"])
    return (f'{{:tx/id {tx[":tx/id"]} :tx/as-of {tx[":tx/as-of"]} '
            f':tx/prev {json.dumps(tx[":tx/prev"])} :tx/cid {json.dumps(tx[":tx/cid"])} '
            f':tx/count {tx[":tx/count"]} :tx/datoms [{datoms}]}}')


def append_tx(tx: dict, log_path: pathlib.Path = LOG_DEFAULT) -> str:
    """Append ONE transaction to the append-only log (never rewrites). Returns the tx CID."""
    log_path.parent.mkdir(parents=True, exist_ok=True)
    if not log_path.exists():
        log_path.write_text(";; watari kotoba Datom log — append-only EAVT transactions "
                            "(content-addressed DAG). Situational-awareness, never surveillance / "
                            "no person-tracking (G2/G4). DO NOT hand-edit. ADR-2606041827.\n",
                            encoding="utf-8")
    with log_path.open("a", encoding="utf-8") as fh:
        fh.write(_tx_to_edn(tx) + "\n")
    return tx[":tx/cid"]


def read_log(log_path: pathlib.Path = LOG_DEFAULT) -> list[dict]:
    """Read the log back as a list of transaction dicts (uses analyze.py's inline EDN reader)."""
    if not log_path.exists():
        return []
    txs = []
    for line in log_path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith(";"):
            continue
        txs.append(_parse(_tokens(line)))
    return txs


def head_cid(log_path: pathlib.Path = LOG_DEFAULT) -> str:
    txs = read_log(log_path)
    return txs[-1][":tx/cid"] if txs else ""


def verify_chain(log_path: pathlib.Path = LOG_DEFAULT) -> dict:
    """Recompute every CID from its datoms + prev; verify the DAG is intact. {ok, length, broken_at}."""
    txs = read_log(log_path)
    prev = ""
    for i, tx in enumerate(txs):
        expect = tx_cid(tx.get(":tx/datoms", []), prev)
        if tx.get(":tx/cid") != expect or tx.get(":tx/prev") != prev:
            return {"ok": False, "length": len(txs), "broken_at": i}
        prev = tx[":tx/cid"]
    return {"ok": True, "length": len(txs), "broken_at": -1}


if __name__ == "__main__":
    import pathlib as _pl
    from analyze import load_edn

    here = _pl.Path(__file__).resolve().parents[1]
    g = here / "data" / "craft-graph.merged.kotoba.edn"
    if not g.exists():
        g = here / "data" / "seed-craft-graph.kotoba.edn"
    datoms = graph_datoms(load_edn(g))
    tx = make_tx(datoms, tx_id=1, as_of=20260608, prev_cid="")
    print(f"# watari kotoba Datom tx — {tx[':tx/count']} EAVT assertions, cid={tx[':tx/cid'][:18]}…")

File: 20-actors/watari/methods/kotoba.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/kotoba.py — the local content-addressed Datom-log writer.
(ns watari.methods.kotoba
  "kotoba.clj — watari kotoba Datom-log writer (local, content-addressed).
  ADR-2606041827 + ADR-2605262130 + ADR-2605312345.

  Canonical state is the kotoba Datom log — content-addressed EAVT assertions, append-only
  (非終末論; the fix stream IS the trajectory, the latest fix IS the current position). This is
  the local, autonomous-loop write path (the shionome/ipaddress/yabai/sukashi/watatsuna shape).

  Constitutional posture holds by construction: outputs are situational-awareness, not
  surveillance / not targeting (G2 — aggregate lane/chokepoint/approach density); a craft is a
  craft, NEVER a person (G4 — no position track linked to an individual, operator is a company
  only). The loop persists exactly what analyze.clj computes, derived signals flagged
  :movement/derived.

    graph-datoms   → EAVT assertions for every entity (craft/fix/leg/lane)
    derived-datoms → EAVT assertions for derived :movement/* signals
    make-tx / append-tx / read-log / head-cid / verify-chain — content-addressed commit-DAG

  EAVT = [op entity attribute value]; op is :db/add only (append-only — no :db/retract).
  Deterministic: caller supplies tx-id + as-of (no wall clock) → resume-safe. The CID is a
  clj-native content address (sha256 over the canonical pr-str of {:datoms :prev}); internally
  consistent + tamper-evident (it does not reproduce kotoba.py's JSON bytes, since the clj port
  carries real EDN keywords where the Python port carried strings)."
  (:require [clojure.java.io :as io]
            [clojure.edn :as edn]
            [clojure.string :as str]))

(def ^:private this-file *file*)
(defn log-default []
  (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile
      (io/file "data" "watari.datoms.kotoba.edn")))

(def id-keys #{:craft/id :craft.fix/id :craft.leg/id :lane/id})

(defn- add-datom [e a v] [:db/add e a v])

(defn graph-datoms
  "Flatten the moving-craft graph into append-only EAVT assertions. E = the entity's id; list
  values fan out. Persists craft identity / fixes / legs / lanes as-is — operator-as-company
  only, never a person (G4)."
  [rows]
  (vec (mapcat
        (fn [r]
          (when (map? r)
            (when-let [e (some #(get r %) id-keys)]
              (for [[k v] r :when (not (id-keys k))
                    item (if (sequential? v) v [v])]
                (add-datom e k item)))))
        rows)))

(defn derived-datoms
  "Flatten the analyzer's derived :movement/* signals into EAVT assertions, each flagged
  :movement/derived true (aggregate situational-awareness, never a per-craft follow/targeting
  feed — G2/G4). `a` is analyze/analyze."
  [{:keys [craft lanes]} a]
  (vec
   (concat
    (mapcat (fn [cp]
              (let [e (str "movement-choke-" cp)]
                [(add-datom e :movement/chokepoint cp)
                 (add-datom e :movement/chokepoint-transit (get-in a [:choke-transit cp]))
                 (add-datom e :movement/derived true)]))
            (sort-by (juxt #(- (get-in a [:choke-transit %])) str) (keys (:choke-transit a))))
    (mapcat (fn [ln]
              (let [e (str "movement-lane-" ln)]
                [(add-datom e :movement/lane ln)
                 (add-datom e :movement/lane-load (get-in a [:lane-load ln]))
                 (add-datom e :movement/vessels (count (get-in a [:lane-kind ln :vessel] #{})))
                 (add-datom e :movement/aircraft (count (get-in a [:lane-kind ln :aircraft] #{})))
                 (add-datom e :movement/derived true)]))
            (sort-by (juxt #(- (get-in a [:lane-load %])) str) (keys (:lane-load a))))
    (mapcat (fn [c]
              (let [fx (get-in a [:latest c])
                    e (str "movement-stale-" c)]
                [(add-datom e :movement/craft c)
                 (add-datom e :movement/stale true)
                 (add-datom e :movement/last-seen (or (:craft.fix/observed-at fx) ""))
                 (add-datom e :movement/derived true)]))
            (sort (:stale a))))))

;; ── content-addressed commit-DAG ──────────────────────────────────────────────
(defn- sha256-hex [^String s]
  (let [md (java.security.MessageDigest/getInstance "SHA-256")]
    (apply str (map #(format "%02x" (bit-and % 0xff)) (.digest md (.getBytes s "UTF-8"))))))

(defn- canonical [datoms prev] (str "{:datoms " (pr-str datoms) " :prev " (pr-str prev) "}"))

(defn tx-cid
  ([datoms] (tx-cid datoms ""))
  ([datoms prev] (str "b" (sha256-hex (canonical datoms prev)))))

(defn make-tx [datoms & {:keys [tx-id as-of prev-cid] :or {prev-cid ""}}]
  {:tx/id tx-id :tx/as-of as-of :tx/prev prev-cid
   :tx/cid (tx-cid datoms prev-cid) :tx/count (count datoms) :tx/datoms datoms})

(defn append-tx
  ([tx] (append-tx tx (log-default)))
  ([tx log-path]
   (let [f (io/file log-path)]
     (.mkdirs (.getParentFile (.getAbsoluteFile f)))
     (when-not (.exists f)
       (spit f (str ";; watari kotoba Datom log — append-only EAVT transactions (content-addressed "
                    "DAG). Situational-awareness, never surveillance / no person-tracking (G2/G4). "
                    "DO NOT hand-edit. ADR-2606041827.\n")))
     (spit f (str (pr-str tx) "\n") :append true)
     (:tx/cid tx))))

(defn read-log
  ([] (read-log (log-default)))
  ([log-path]
   (let [f (io/file log-path)]
     (if-not (.exists f)
       []
       (->> (str/split-lines (slurp f))
            (map str/trim)
            (remove #(or (empty? %) (str/starts-with? % ";")))
            (mapv edn/read-string))))))

(defn head-cid
  ([] (head-cid (log-default)))
  ([log-path] (let [txs (read-log log-path)] (if (seq txs) (:tx/cid (last txs)) ""))))

(defn verify-chain
  ([] (verify-chain (log-default)))
  ([log-path]
   (let [txs (read-log log-path)]
     (loop [i 0 prev "" xs txs]
       (if (empty? xs)
         {:ok true :length (count txs) :broken-at -1}
         (let [tx (first xs)
               expect (tx-cid (:tx/datoms tx []) prev)]
           (if (or (not= (:tx/cid tx) expect) (not= (:tx/prev tx) prev))
             {:ok false :length (count txs) :broken-at i}
             (recur (inc i) (:tx/cid tx) (rest xs)))))))))
